spark/src/test/java/org/apache/iceberg/TestRemoveOrphanFilesAction.java

```diff
@@ -0,0 +1,327 @@
+
+    List<ThreeColumnRecord> records = Lists.newArrayList(
+        new ThreeColumnRecord(1, "AAAAAAAAAA", "AAAA")
+    );
+
+    Dataset<Row> df = spark.createDataFrame(records, ThreeColumnRecord.class).coalesce(1);
+
+    df.select("c1", "c2", "c3")
+        .write()
+        .format("iceberg")
+        .mode("append")
+        .save(tableLocation);
+
+    df.select("c1", "c2", "c3")
+        .write()
+        .format("iceberg")
+        .mode("append")
+        .save(tableLocation);
+
+    List<String> validFiles = spark.read().format("iceberg")
+        .load(tableLocation + "#files")
+        .select("file_path")
+        .as(Encoders.STRING())
+        .collectAsList();
+    Assert.assertEquals("Should be 2 valid files", 2, validFiles.size());
+
+    df.write().mode("append").parquet(tableLocation + "/data");
+
+    Path dataPath = new Path(tableLocation + "/data");
+    FileSystem fs = dataPath.getFileSystem(spark.sessionState().newHadoopConf());
+    List<String> allFiles = Arrays.stream(fs.listStatus(dataPath, HiddenPathFilter.get()))
+        .filter(FileStatus::isFile)
+        .map(file -> file.getPath().toString())
+        .collect(Collectors.toList());
+    Assert.assertEquals("Should be 3 files", 3, allFiles.size());
+
+    List<String> invalidFiles = Lists.newArrayList(allFiles);
+    invalidFiles.removeAll(validFiles);
+    Assert.assertEquals("Should be 1 invalid file", 1, invalidFiles.size());
+
+    // sleep for 1 second to unsure files will be old enough
+    Thread.sleep(1000);
+
+    Actions actions = Actions.forTable(table);
+
+    RemoveOrphanFilesActionResult result1 = actions.removeOrphanFiles()
+        .allDataFilesTable(tableLocation + "#all_data_files")
+        .olderThan(System.currentTimeMillis())
+        .dryRun(true)
+        .execute();
+    Assert.assertEquals("Action should find 1 data file", invalidFiles, result1.dataFiles());
+    Assert.assertTrue("Invalid file should be present", fs.exists(new Path(invalidFiles.get(0))));
+
+    RemoveOrphanFilesActionResult result2 = actions.removeOrphanFiles()
+        .allDataFilesTable(tableLocation + "#all_data_files")
+        .olderThan(System.currentTimeMillis())
+        .execute();
+    Assert.assertEquals("Action should delete 1 data file", invalidFiles, result2.dataFiles());
+    Assert.assertFalse("Invalid file should not be present", fs.exists(new Path(invalidFiles.get(0))));
+
+    List<ThreeColumnRecord> expectedRecords = Lists.newArrayList();
+    expectedRecords.addAll(records);
+    expectedRecords.addAll(records);
+
+    Dataset<Row> resultDF = spark.read().format("iceberg").load(tableLocation);
+    List<ThreeColumnRecord> actualRecords = resultDF
+        .as(Encoders.bean(ThreeColumnRecord.class))
+        .collectAsList();
+    Assert.assertEquals("Rows must match", expectedRecords, actualRecords);
+  }
+
+  @Test
+  public void testAllValidFilesAreKept() throws IOException, InterruptedException {
```

Comment:
Can you also add a test for Write-Audit-Publish (WAP) workflow case where a snapshot can be staged (using the cherrypicking operation), where it's not part of the list of active snapshots. So expected behavior should be that this action should not delete those staged files as orphan files.
There are tests in `TestWapWorkflow` that illustrate this case.

Reply:
Good idea, will add a case for that.

Reply:
I agree it would be nice to have a test for it. This should work because the metadata tables used return all files reachable by metadata, not just the ones in a single snapshot. We use the same table for a similar check in our environment.

Reply:
There is one case when `all_data_file_table` won't report WAP files: when there is no snapshot.

```
  @Override
  public CloseableIterable<FileScanTask> planFiles() {
    Snapshot snapshot = snapshot();
    if (snapshot != null) {
      LOG.info("Scanning table {} snapshot {} created at {} with filter {}", table,
          snapshot.snapshotId(), formatTimestampMillis(snapshot.timestampMillis()),
          rowFilter);

      Listeners.notifyAll(
          new ScanEvent(table.toString(), snapshot.snapshotId(), rowFilter, schema()));

      return planFiles(ops, snapshot, rowFilter, caseSensitive, colStats);

    } else {
      LOG.info("Scanning empty table {}", table);
      return CloseableIterable.empty();
    }
  } 
```

Reply:
Also, `table.currentSnapshot().manifestListLocation()` in `location()`can lead to a NPE.

Reply:
I've created #904 and #905 to address later.

Reply:
As long as we have a current snapshot, it does work correctly. I've added a test.

Reply:
I thought we addressed the metadata table problem when there is no current snapshot in #801. I'll check to see why that doesn't work. My initial guess is that PR refers to static tables and this is a parallel table.

Reply:
It's just individual tables that were fixed. The solution is to override `planFiles`:

```java
    @Override
    public CloseableIterable<FileScanTask> planFiles() {
      // override planFiles to avoid the check for a current snapshot because this metadata table is for all snapshots
      return CloseableIterable.withNoopClose(HistoryTable.this.task(this));
    }
```